### ISSUE TITLE
[suggestion] Provide example CI fuzzing configs

### ISSUE THREAD

**Issue by terrelln** [google/oss-fuzz#2858]
First, thanks for oss-fuzz! It is a huge step up from our existing fuzzing infra for zstd. We find subtle bugs much quicker than before because of the continuous builds, and more CPU.

Zstd runs regression tests on the current corpora in its CI that gets run on every PR. I would love to fuzz for 5-10 minutes on every PR, starting from the current corpus. Is this possible, and could we add example configs for popular CIs to oss-fuzz?

I might find some time to look into it at some point, but if there was an easy example that I could copy I would definitely do it.

**Comment by jonathanmetzman** [comment]
>First, thanks for oss-fuzz! It is a huge step up from our existing fuzzing infra for zstd. We find subtle bugs much quicker than before because of the continuous builds, and more CPU.

Your welcome! :-)

>Zstd runs regression tests on the current corpora in its CI that gets run on every PR. I would love to fuzz for 5-10 minutes on every PR, starting from the current corpus. Is this possible, and could we add example configs for popular CIs to oss-fuzz?

I have plans to implement something like this soon. Stay tuned.

Question for you: What are the different CI systems zstd uses for? Is there one that is a better fit for this sort of thing?

**Comment by terrelln** [comment]
We use both Travis-CI and CircleCI. Either would be a good for us. Our only constraint is that we we don't want to spend more than ~30 minutes for any pull request CI run, but we can run all the fuzzers in parallel.

**Comment by jonathanmetzman** [comment]
Thanks, this helps, I'm still planning this out right now so I'm not exactly sure what shape it will take.

**Comment by evverx** [comment]
Looks like this issue is a variation of https://github.com/google/oss-fuzz/issues/1788 and one possible solution would be the webhook (or something like that) we discussed in https://github.com/google/oss-fuzz/issues/2525#issuecomment-502927581.

**Comment by evverx** [comment]
@terrelln just out of curiosity, does the zstd project use the free Circle CI offering? Last time I looked at it it couldn't even withstand the unit test run by the systemd project let alone fuzzing when PRs are opened.

**Comment by terrelln** [comment]
I believe we are using the free CircleCI. I haven't had any problems with memory/cpu on the free CircleCI, but maybe fuzzing would expose some.

**Comment by jonathanmetzman** [comment]
> Looks like this issue is a variation of #1788 and one possible solution would be the webhook (or something like that) we discussed in [#2525 (comment)](https://github.com/google/oss-fuzz/issues/2525#issuecomment-502927581).

Right. I'll close this then as a duplicate.

**Comment by evverx** [comment]
It would be interesting to figure out whether fuzzing for several minutes is better than just running fuzzers against the corpora updated on a regular basis (because in terms of resources it's more expensive). In `systemd` the only bug that was missed in a PR and could have been caught by additional fuzzing was https://github.com/systemd/systemd/issues/13591.

@jonathanmetzman speaking of https://github.com/systemd/systemd/issues/13591, is the "alignment" check off on OSS-Fuzz for the same reasons `pointer-overflow` is off there (which we discussed in https://github.com/google/oss-fuzz/pull/2701#issuecomment-533735673)?

**Comment by jonathanmetzman** [comment]
> It would be interesting to figure out whether fuzzing for several minutes is better than just running fuzzers against the corpora updated on a regular basis (because in terms of resources it's more expensive). In `systemd` the only bug that was missed in a PR and could have been caught by additional fuzzing was [systemd/systemd#13591](https://github.com/systemd/systemd/issues/13591).
> 

I suspect PR fuzzing could do a lot. In Chrome we once did some quick queries and we found that roughly half of bugs are found the day after they are introduced. I think PR fuzzing could probably find a large fraction of this half but I doubt it could replace the current model. I agree it's probably more "efficient" though.

> @jonathanmetzman speaking of [systemd/systemd#13591](https://github.com/systemd/systemd/issues/13591), is the "alignment" check off on OSS-Fuzz for the same reasons `pointer-overflow` is off there (which we discussed in [#2701 (comment)](https://github.com/google/oss-fuzz/pull/2701#issuecomment-533735673))?

I'm not sure to be honest why the alignment check isn't on. As far as I can tell we haven't tested it yet and gotten any feedback. More likely, enabling it is just a lower priority than our other work, since we would have to deal with complaints about false/useless positives etc.

**Comment by jonathanmetzman** [comment]
> I suspect PR fuzzing could do a lot. In Chrome we once did some quick queries and we found that roughly half of bugs are found the day after they are introduced. I think PR fuzzing could probably find a large fraction of this half but I doubt it could replace the current model. I agree it's probably more "efficient" though.
> 

Maybe I misunderstood this. Were you suggesting just running against corpus in CI? Or running against corpus+fuzzing in CI?

**Comment by evverx** [comment]
I kind of tried to compare running fuzz targets against the corpora with fuzzing for several minutes starting with the corpora when PRs are opened (in CI). In `systemd` the first approach is used and so far we've caught everything except for that one bug before PRs were merged. If we had used the second approach we would have caught all of them.

If resources isn't an issue it's certainly better to fuzz (especially when new fuzz targets are introduced) but running the fuzzers against the corpora seems to be good enough in our particular case.

Just to clarify, I'm not advocating for replacing continuous fuzzing with fuzzing when PRs are opened only. After all, among other things continuous fuzzing is how the corpora are updated on a regular basis :-)

**Comment by jonathanmetzman** [comment]
> I kind of tried to compare running fuzz targets against the corpora with fuzzing for several minutes starting with the corpora when PRs are opened (in CI). In `systemd` the first approach is used and so far we've caught everything except for that one bug before PRs were merged. If we had used the second approach we would have caught all of them.
> 
> If resources isn't an issue it's certainly better to fuzz (especially when new fuzz targets are introduced) but running the fuzzers against the corpora seems to be good enough in our particular case.
> 
> Just to clarify, I'm not advocating for replacing continuous fuzzing with fuzzing when PRs are opened only. After all, among other things continuous fuzzing is how the corpora are updated on a regular basis :-)

Right. This raises two interesting points:
1. If fuzzing is run on arbitrary pull requests and uses the current corpus, isn't there a danger that this can be leaked by a malicious pull request? My solution to this is using a corpus that is 90 days old (ie: won't have any non-disclosed bugs in it, because of our policy).

2. If the corpus is being updated, is it possible bugs can be found on a later pull request rather than the correct one. There are other things that could cause this but I'm basically curious how you've dealt with this sort of nondeterminism in finding bugs. Has finding bugs in a later PR been an issue for you?

**Comment by evverx** [comment]
> If fuzzing is run on arbitrary pull requests and uses the current corpus, isn't there a danger that this can be leaked by a malicious pull request?

We keep our corpora "in the cloud" only four people have access to. The fuzzers are sent there and run there. We never pull the corpora when PRs are opened, which, as far as I can tell, helps to avoid the issue. That's basically why I said that it would be great to bring fuzz targets to OSS-Fuzz instead of downloading them from GCP.

> If the corpus is being updated, is it possible bugs can be found on a later pull request rather than the correct one?

Testcases crashing our fuzzers are automatically excluded from the corpora so "permanent" bugs don't affect our CI. It's of course possible that at some point the corpora and PRs won't be in sync but we don't expect it to happen very often mainly because we turned this kind of CI after  about a year of continuous fuzzing on OSS-Fuzz when I guess all the shallow bugs and most not-so-shallow bugs had been fixed.
